src/sentry/monitors/processing_errors.py

```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
     """This monitor environment is sending checkins too frequently"""
     ORGANIZATION_KILLSWITCH_ENABLED = 15
     """We have disabled checkin ingestion for this org. Contact support for details"""
+
+
+class CheckinValidationError(Exception):
+    def __init__(self, processing_errors: list[ProcessingError], monitor: Monitor | None = None):
+        # Monitor is optional, since we don't always have the monitor related to the checkin available
+        self.processing_errors = processing_errors
+        self.monitor = monitor
+
+
+class ProcessingErrorData(TypedDict):
+    type: str
+    data: dict[str, Any]
+
+
+@dataclasses.dataclass(frozen=True)
+class ProcessingError:
+    type: ProcessingErrorType
+    data: dict[str, Any] = dataclasses.field(default_factory=dict)
+
+    def to_dict(self) -> ProcessingErrorData:
+        return {
+            "type": self.type.name,
+            "data": self.data,
+        }
+
+    @classmethod
+    def from_dict(cls, processing_error_data: ProcessingErrorData) -> ProcessingError:
+        return cls(
+            ProcessingErrorType[processing_error_data["type"]],
+            processing_error_data["data"],
+        )
+
+
+class CheckinProcessingErrorData(TypedDict):
+    errors: list[ProcessingErrorData]
+    checkin: CheckinItemData
+
+
+@dataclasses.dataclass(frozen=True)
+class CheckinProcessingError:
+    errors: list[ProcessingError]
+    checkin: CheckinItem
+
+    def to_dict(self) -> CheckinProcessingErrorData:
+        return {
+            "errors": [error.to_dict() for error in self.errors],
+            "checkin": self.checkin.to_dict(),
+        }
+
+    @classmethod
+    def from_dict(cls, data: CheckinProcessingErrorData) -> CheckinProcessingError:
+        return cls(
+            errors=[ProcessingError.from_dict(error) for error in data["errors"]],
+            checkin=CheckinItem.from_dict(data["checkin"]),
+        )
+
+
+class CheckinProcessErrorsManager:
+    def _get_cluster(self):
+        return redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+
+    def store(self, error: CheckinProcessingError, monitor: Monitor | None):
+        if monitor is None:
+            # Attempt to get the monitor from the checkin info if we failed to retrieve it during ingestion
+            try:
+                monitor = Monitor.objects.get(
+                    project_id=error.checkin.message["project_id"],
+                    slug=error.checkin.payload["monitor_slug"],
+                )
+            except Monitor.DoesNotExist:
+                pass
+        if monitor:
+            error_identifier = f"monitor:{monitor.id}"
+        else:
+            error_identifier = f'project:{error.checkin.message["project_id"]}'
```

Comment:
wish we could use `build_project_identifier`. Tempted to say the signature could be `project_or_project_id` but feels kinda gross